src/components/implementation/cbuf_mgr/default/cbuf_mgr.c

```diff
@@ -190,86 +215,148 @@ cbuf_comp_info_bin_get(struct cbuf_comp_info *cci, int sz)
+	if (!cbuf_map(spdid, dest, new_p, sz, flags)) goto done;
 
-free:
+free: 
 	if (dest) valloc_free(cos_spd_id(), spdid, (void *)dest, 1);
-	dest = 0;
-	page_free(p, size/PAGE_SIZE);
-	p = 0;
+	if (!p) page_free(new_p, sz/PAGE_SIZE);
 	ret = -1;
-done:
-	*page  = p;
+
+done:	
+	if (page) *page  = new_p;
 	*daddr = dest;
+	rdtscll(end);
+	op_nums[7]++;
+	per_total[7] += (end-start);
 	return ret;
 }
 
-/* Do any components have a reference to the cbuf? */
+/* Do any components have a reference to the cbuf? 
+ * key function coordinates manager and client.
+ * When this returns 1, this cbuf may or may not be used by some components.
+ * When this returns 0, it guarantees: 
+ * If all clients use the protocol correctly, there is no reference 
+ * to the cbuf and no one will receive the cbuf after this. Furthermore, 
+ * if the cbuf is in some free list, its inconsistent bit is already set.
+ * That is to say, the manager can safely collect or re-balance this cbuf.
+ *
+ * Proof: 1. If a component gets the cbuf from free-list, it will 
+ * simply discard this cbuf as its inconsistent bit is set.
+ * 2. Assume component c sends the cbuf. 
+ * It is impossible to send the cbuf after we check c's refcnt, since c 
+ * has no reference to this cbuf.
+ * If this send happens before we check c's refcnt, because the sum of 
+ * nsent is equal to the sum of nrecv, this send has been received and 
+ * no further receive will happen.
+ * 3. No one will call cbuf2buf to receive this cbuf after this function, 
+ * as all sends have been received and no more sends will occur during this function
+ *
+ * However, if clients do not use protocol correctly, this function 
+ * provides no guarantee. cbuf_unmap_prepare takes care of this case.
+ */
 static int
 cbuf_referenced(struct cbuf_info *cbi)
 {
 	struct cbuf_maps *m = &cbi->owner;
-	int sent, recvd;
+	int sent, recvd, ret = 1;
+	unsigned long old_nfo, new_nfo;
+	unsigned long long old;
+	struct cbuf_meta *mt, *own_mt = m->m;
+
+	old_nfo = own_mt->nfo;
+	new_nfo = old_nfo | CBUF_INCONSISENT;
+	if (unlikely(!cos_cas(&own_mt->nfo, old_nfo, new_nfo))) {
+		goto done;
+	}
 
+	mt   = (struct cbuf_meta *)(&old);
 	sent = recvd = 0;
 	do {
 		struct cbuf_meta *meta = m->m;
 
-		if (meta) {
-			if (CBUF_REFCNT(meta)) return 1;
-			sent  += meta->snd_rcv.nsent;
-			recvd += meta->snd_rcv.nrecvd;
+		/* Guarantee atomically read the two words (refcnt and nsent/nrecv).
+		 * Consider this case, c0 sends a cbuf to c1 and frees this 
+		 * this cbuf, but before c1 receives it, the manager comes in 
+		 * and checks c1's refcnt. Now it is zero. But before the manager 
+		 * checks c1's nsent/nrecv, it is preempted by c1. c1 receives 
+		 * this cbuf--increment refcnt and nsent/nrecv. After this, we 
+		 * switch back the manager, who will continues to check c1's 
+		 * nsent/nrecv, now it is 1, which is equal to c0's nsent/nrecv. 
+		 * Thus the manager can collect or unmap this cbuf.*/
+		old = *(unsigned long long *)meta;
```

Comment:
This is a bit hackish and works because meta->nfo and meta->snd_rcv are within the same 64-bits in the `struct meta`. Is this implicit requirement checked or enforced anywhere? What prevents someone from adding another field to `struct meta` that would break this code?